### PR TITLE
feat: Add support for GCP Secret Manager

### PR DESCRIPTION
**Pull request by loxley**
### Description
Add GCP Secret Manager as a new backend

### Checklist
Please make sure that your PR fulfills the following requirements:
- [X] Reviewed the guidelines for contributing to this repository
- [X] The commit message follows the [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [X] Tests for the changes have been updated
- [X] Docs have been added / updated

### Type of Change

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [ ] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

### CODE REVIEW

**Comment by werne2j** [comment]
@loxley Thank you for the PR! Will try to take a look tomorrow!

**Comment by werne2j** [comment]
@loxley could you please add a test for the new config to `config_test.go`?

File: docs/backends.md [hunk 1]
```diff
@@ -114,3 +114,13 @@ These are the parameters for AWS:
...
+
+### GCP Secret Manager
+
+#### GCP Authentication
+Refer to the [Authentication Overview](https://cloud.google.com/docs/authentication) for Google Cloud APIs.
+
+These are the parameters for GCP:
+```
+AVP_TYPE: gcpsecretmanager
+```
```
Comment (werne2j) [inline docs/backends.md:126]: Could you add an example yaml to the GCP docs section? To show an example of what an annotation or inline path would look like?

Comment (josandstrom) [reply docs/backends.md:126]: Sure! Will fix! :+1:

**Comment by josandstrom** [comment]
@werne2j added test and updated docs.

**Comment by jkayani** [comment]
This looks great, thanks for the work! A few failing tests to fix:

```
ok  	github.com/IBM/argocd-vault-plugin/pkg/backends	7.662s	coverage: 79.0% of statements
=== RUN   TestNewConfig
    config_test.go:104: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
--- FAIL: TestNewConfig (0.00s)
=== RUN   TestNewConfigNoType
    config_test.go:123: expected error Must provide a supported Vault Type to be thrown, got google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
--- FAIL: TestNewConfigNoType (0.00s)
```

Looks like the SDK is expecting some default credentials - for the purpose of testing, maybe there is a dummy value we can set, like what is done here: https://github.com/IBM/argocd-vault-plugin/blob/main/pkg/config/config_test.go#L284?

**Comment by loxley** [comment]
> Looks like the SDK is expecting some default credentials - for the purpose of testing, maybe there is a dummy value we can set, like what is done here: https://github.com/IBM/argocd-vault-plugin/blob/main/pkg/config/config_test.go#L284?

Right! Let me take a look!

File: docs/backends.md [hunk 2]
```diff
@@ -114,3 +114,24 @@ These are the parameters for AWS:
 AVP_TYPE: awssecretsmanager
 AWS_REGION: Your AWS Region (Optional: defaults to us-east-2)
 ```
+
+### GCP Secret Manager
+
+#### GCP Authentication
```
Comment (werne2j) [inline docs/backends.md:120]: Sorry, one more small nitpick with this doc. Can you update the docs to look like this:

### GCP Secret Manager

##### GCP Authentication
Refer to the [Authentication Overview](https://cloud.google.com/docs/authentication) for Google Cloud APIs.

These are the parameters for GCP:
```
AVP_TYPE: gcpsecretmanager
```

##### Examples

###### Path Annotation
```yaml
kind: Secret
apiVersion: v1
metadata:
  name: test-secret
  annotations:
    avp.kubernetes.io/path: projects/12345678987/secrets/test-secret/versions/latest
type: Opaque
data:
  password: <test-secret>
```

###### Inline Path
```yaml
kind: Secret
apiVersion: v1
metadata:
  name: test-secret
type: Opaque
data:
  password: <path:projects/12345678987/secrets/test-secret/versions/latest#test-secret>
```

Thanks!

**Comment by loxley** [comment]
@jkayani where would you want me to put a dummy json file that will be used to solve the default application credentials issue?

**Comment by jkayani** [comment]
@loxley There is a `fixtures/input` folder it can go in: https://github.com/IBM/argocd-vault-plugin/tree/main/fixtures/input

**Comment by loxley** [comment]
OK, pushed changes to docs and default application credentials. Hope its better now! :)

**Comment by werne2j** [comment]
@loxley the code looks good with a quick glance, can you resolve the conflicts and we will go ahead and finish the review

**Comment by loxley** [comment]
> @loxley the code looks good with a quick glance, can you resolve the conflicts and we will go ahead and finish the review

Done!

**Review by werne2j — APPROVED** [review]
Looks good! Thanks @loxley!

**Review by jkayani — APPROVED** [review]

**Pull request merged**